### PR TITLE
Add functions to copy data from xarray to TileDB

### PR DESCRIPTION
**Pull request by jp-dark**

### CODE REVIEW

File: tiledb/cf/xarray_engine/array_wrapper.py [hunk 1]
```diff
@@ -209,4 +210,4 @@ def set_metadata(self, input_meta):
         key_prefix = f"{_ATTR_PREFIX}{self._attr_name}"
         with tiledb.open(**self._array_kwargs, mode="w") as array:
             for key, value in input_meta.items():
-                array.meta[f"{key_prefix}.{key}"] = value
+                safe_set_metadata(array.meta, f"{key_prefix}.{key}", value)
```
Comment (perrygeo) [inline tiledb/cf/xarray_engine/array_wrapper.py:213]: Under what conditions would the direct metadata assignment fail? Is it ok to ignore those and proceed with only a warning?

File: tiledb/cf/xarray_engine/_encoding.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+            # Set attribute encodings: attr_name, attr_filters, and attr_fill.
+            self.attr_name = encoding.get(
+                _ATTR_NAME_ENCODING,
+                f"{self._name}_" if self._name in variable.dims else self._name,
+            )
+            self.attr_filters = encoding.get(
+                _ATTR_FILTERS_ENCODING,
+                tiledb.FilterList(
+                    (tiledb.ZstdFilter(level=5, ctx=self._ctx),), ctx=self._ctx
+                ),
```
Comment (jp-dark) [inline tiledb/cf/xarray_engine/_encoding.py:63]: Add bit shuffle as well? (Would match default for NetCDF).

File: tiledb/cf/xarray_engine/_encoding.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+            )
+            self.attr_fill = encoding.get(
+                _FILL_VALUE_ENCODING,
+                self._variable.encoding.get(_FILL_VALUE_ENCODING, None),
+            )
```
Comment (jp-dark) [inline tiledb/cf/xarray_engine/_encoding.py]: Over-riding might cause more problems than it is worth. Maybe remove this and just always use the variable fill instead.

File: tiledb/cf/xarray_engine/_encoding.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+    @property
+    def encoding(self):
+        return self._encoding
+
+    @property
+    def tiles(self):
+        return self._encoding[_TILE_SIZES_ENCODING]
+
+    @tiles.setter
```
Comment (jp-dark) [inline tiledb/cf/xarray_engine/_encoding.py:205]: Should I add a warning if the tile sizes are very large or very small? Not sure if a warning would be more annoying or helpful.

File: tiledb/cf/xarray_engine/_encoding.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+                    "Encoding error on variable '{self._name}'. Invalid encoding key "
+                    f"``{key}``."
+                )
+
+        # Initialize encoding values.
+        try:
+            # Set attribute encodings: attr_name, attr_filters, and attr_fill.
+            self.attr_name = encoding.get(
+                _ATTR_NAME_ENCODING,
+                f"{self._name}_" if self._name in variable.dims else self._name,
```
Comment (jp-dark) [inline tiledb/cf/xarray_engine/_encoding.py:57]: Coordinates in xarray generally have matching variable and dimension names. The convention here is just adding an `_` after the variable name. Is there a better choice?

File: tests/xarray_engine/test_from_xarray.py [hunk 6]
```diff
@@ -0,0 +1,522 @@
...
+    name = "simple_dask"
+
+    kwargs = {"encoding": {"example": {"tiles": (8, 8)}}}
+
+    ds = xr.Dataset(
+        {
+            "example": xr.DataArray(
+                da.from_array(
+                    np.reshape(np.arange(1024, dtype=np.float32), (32, 32)),
+                    chunks=((8, 16, 8), (16, 16)),
```
Comment (perrygeo) [inline tests/xarray_engine/test_from_xarray.py:427]: idea: check number of fragments and tile alignment

File: tiledb/cf/_utils.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    elif isinstance(value, np.generic):
+        value = (value.tolist(),)
+    try:
+        meta[key] = value
+    except ValueError as err:  # pragma: no cover
+        with warnings.catch_warnings():
+            warnings.warn(
+                f"Failed to set metadata `{key}={value}` with error: {err}",
+                stacklevel=3,
+            )
```
Comment (perrygeo) [inline tiledb/cf/_utils.py]:
```suggestion
    meta[key] = value
```

File: tiledb/cf/xarray_engine/_encoding.py (hunk 2 above)
Comment (perrygeo) [reply tiledb/cf/xarray_engine/_encoding.py:63]: Allow default filter to be passed for all writes?

File: tiledb/cf/xarray_engine/_encoding.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+        return self._encoding.get(_ATTR_NAME_ENCODING, self._name)
+
+    @attr_name.setter
+    def attr_name(self, name):
+        if name in self._variable.dims:
+            raise ValueError(
+                f"Encoding error for variable '{self._name}'. Attribute name "
+                f"'{name}' is already a dimension name. Set a different attribute "
+                f"name using the ``{_ATTR_NAME_ENCODING}`` encoding key."
+            )
```
Comment (perrygeo) [inline tiledb/cf/xarray_engine/_encoding.py]: jpd: edit error message

File: tiledb/cf/xarray_engine/_encoding.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+                # remaining dimensions to the size of the variable dimension.
+                unlimited_dims = {
+                    dim_name for dim_name in variable.dims if dim_name in unlimited_dims
+                }
+                unlim = np.iinfo(self.dim_dtype).max
+                self.max_shape = tuple(
+                    unlim if dim_name in unlimited_dims else dim_size
+                    for dim_name, dim_size in zip(variable.dims, variable.shape)
+                )
+            self.tiles = encoding.get(_TILE_SIZES_ENCODING, None)
```
Comment (perrygeo) [inline tiledb/cf/xarray_engine/_encoding.py:81]: If None (default) sanity check the tile sizes and warn if the product of tile sizes * dtype size > threshold (1 GB?)

File: tiledb/cf/xarray_engine/writer.py [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+    region: Optional[Mapping[str, slice]],
+    copy_group_metadata: bool,
+    copy_variable_metadata: bool,
+    copy_variable_data: bool,
+):
+    """Copies data and metadata from an xarray dataset to a TileDB group corresponding
+    to the dataset.
+
+    Optionally copies metadata as well as variable data.
+
```
Comment (perrygeo) [inline tiledb/cf/xarray_engine/writer.py]:
```suggestion
Parameters
---------------
```

File: tiledb/cf/xarray_engine/writer.py [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+            raise ValueError(
+                f"``region`` contains a slice with step={dim_slice.step} on dimension "
+                f"'{dim_name}. All slices must have step size 1 or None."
+            )
+
+    # Copy group metadata
+    if copy_group_metadata:
+        with tiledb.Group(group_uri, mode="w", config=config, ctx=ctx) as group:
+            for key, val in group_metadata:
+                group.meta[key] = val
```
Comment (perrygeo) [inline tiledb/cf/xarray_engine/writer.py]: use safe metadata set method

File: tiledb/cf/xarray_engine/writer.py [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+            )
+
+    return variables, group_metadata
+
+
+def get_chunk_regions(
+    target_region: Tuple[slice, ...],
+    source_shape: Tuple[int, ...],
+    chunks: Optional[Tuple[Tuple[int, ...], ...]],
+):
```
Comment (perrygeo) [inline tiledb/cf/xarray_engine/writer.py]: Consider specifying the return type as generator of tuples

File: tests/xarray_engine/test_from_xarray.py [hunk 13]
```diff
@@ -0,0 +1,522 @@
...
+
+    @pytest.fixture(scope="class")
+    def expected_schemas(self):
+        default_filters = tiledb.FilterList([tiledb.ZstdFilter(level=5)])
+        return {
+            "example": tiledb.ArraySchema(
+                tiledb.Domain(tiledb.Dim("x", domain=(0, 15), tile=8, dtype=np.uint32)),
+                [tiledb.Attr("example", filters=default_filters, dtype=np.int32)],
+            )
+        }
```
Comment (perrygeo) [inline tests/xarray_engine/test_from_xarray.py:522]: Consider adding test for GDAL schema rasters. 

```
    s = tiledb.ArraySchema(
        domain=tiledb.Domain(
            tiledb.Dim(
                name="BANDS", domain=(0, len(BANDS) - 1), tile=1, dtype=np.uint64
            ),
            tiledb.Dim(name="Y", domain=(0, height), tile=1024, dtype=np.uint64),
            tiledb.Dim(name="X", domain=(0, width), tile=1024, dtype=np.uint64),
        ),
        attrs=[
            tiledb.Attr(name="TDB_VALUES", dtype=np.uint16, filters=filters),
        ],
        cell_order="row-major",
        tile_order="row-major",
        sparse=False,
        allows_duplicates=False,
    )
```
+ GDAL's metadata

**Review by perrygeo — APPROVED** [review]
Comments are suggestions, no blockers.

File: tiledb/cf/xarray_engine/_encoding.py (hunk 9 above)
Comment (perrygeo) [reply tiledb/cf/xarray_engine/_encoding.py:81]: Also, sanity check for the presence of chunk sizes that may imply a "natural" tile size.

File: tiledb/cf/xarray_engine/_encoding.py (hunk 2 above)
Comment (jp-dark) [reply tiledb/cf/xarray_engine/_encoding.py:63]: https://app.shortcut.com/tiledb-inc/story/32780

File: tiledb/cf/xarray_engine/_encoding.py (hunk 4 above)
Comment (jp-dark) [reply tiledb/cf/xarray_engine/_encoding.py:205]: https://app.shortcut.com/tiledb-inc/story/32781

File: tiledb/cf/xarray_engine/_encoding.py (hunk 9 above)
Comment (jp-dark) [reply tiledb/cf/xarray_engine/_encoding.py:81]: https://app.shortcut.com/tiledb-inc/story/32781

File: tests/xarray_engine/test_from_xarray.py (hunk 13 above)
Comment (jp-dark) [reply tests/xarray_engine/test_from_xarray.py:522]: https://app.shortcut.com/tiledb-inc/story/32782

File: tests/xarray_engine/test_from_xarray.py (hunk 6 above)
Comment (jp-dark) [reply tests/xarray_engine/test_from_xarray.py:427]: https://app.shortcut.com/tiledb-inc/story/32784/test-number-of-fragments-create-by-xarray-writer

**Comment by itsgifnotjiff** [comment]
I love Mrs. @jp-dark !!! Thank you so much and I hope the people at Pangeo are paying attention :)

**Pull request merged**